lib/tests/unit/connector/connector_test.cc

```diff
@@ -157,6 +160,115 @@ TEST_CASE("Connector::startMonitoring", "[connector]") {
 
         REQUIRE(num_connections > 1);
         REQUIRE(num_pings > 0);
+
+        REQUIRE_NOTHROW(c.stopMonitoring());
+    }
+}
+
+TEST_CASE("Connector::monitorConnection", "[connector]") {
+    SECTION("reconnects after the broker stops") {
+        std::unique_ptr<Connector> c_ptr;
+        uint16_t port;
+        Util::thread monitor;
+
+        {
+            MockServer mock_server;
+            bool connected = false;
+            mock_server.set_open_handler(
+                    [&connected](websocketpp::connection_hdl hdl) {
+                        connected = true;
+                    });
+            mock_server.go();
+            port = mock_server.port();
+
+            c_ptr.reset(new Connector("wss://localhost:" + std::to_string(port) + "/pcp",
+                                      "test_client",
+                                      getCaPath(), getCertPath(), getKeyPath(),
+                                      WS_TIMEOUT_MS, ASSOCIATION_TIMEOUT_S,
+                                      ASSOCIATION_REQUEST_TTL_S,
+                                      PONG_TIMEOUTS_BEFORE_RETRY));
+
+            REQUIRE_FALSE(connected);
+            REQUIRE_NOTHROW(c_ptr->connect(1));
+            REQUIRE(c_ptr->isConnected());
+            REQUIRE(connected);
+
+            // Infinite retries; 1 s between each WebSocket ping
+            monitor = Util::thread { [&]() {
+                REQUIRE_NOTHROW(c_ptr->monitorConnection(0, 1));
+            } };
+        }
+
+        // The broker does not exist anymore (RAII)
+        wait_for([&c_ptr](){return !c_ptr->isConnected();});
+
+        REQUIRE_FALSE(c_ptr->isConnected());
+
+        MockServer mock_server {port};
+        mock_server.go();
+
+        wait_for([&c_ptr](){return c_ptr->isConnected();});
+
+        REQUIRE(c_ptr->isConnected());
+
+        wait_for([&c_ptr](){return c_ptr->isAssociated();});
+
+        REQUIRE(c_ptr->isAssociated());
+
+        REQUIRE_NOTHROW(c_ptr->stopMonitoring());
+        // Ensure monitoring halts completely before we attempt to destroy the connection.
+        // Otherwise destruction will attempt to cleanup the object before startMonitorTask
+        // has exited.
+        monitor.join();
+    }
+
+    SECTION("reconnects after 1 pong timeout") {
```

Comment:
It would be nice to test the store / throw mechanism for exceptions. I'll take care of that.